### PR TITLE
[Python] DECIMAL with value 0.00... issue fix

### LINKED ISSUE #4688: Regression in Decimal handling in 0.5.0
### What happens?

We are encountering a regression with the handling of dataframes with Python `Decimal` values in them with the 0.5.0 release. It appears to be an issue with how the width and scale parameters are inferred. I have checked our minimal repro and 0.4.0 is not affected, but 0.5.0 and the latest master branch builds are.

### To Reproduce

First we construct a dataframe with some Decimal values in it:

```python
import pandas as pd
from decimal import Decimal

decimals = pd.DataFrame(
    data={
        "decimals": [
            Decimal("0.00"),
            Decimal("125.90"),
            Decimal("0.00"),
            Decimal("2502.63"),
            Decimal("0.00"),
            Decimal("0.00"),
            Decimal("321.00"),
        ]
    }
)
```

Then we try to query it:

```python
import duckdb

conn = duckdb.connect(database= ":memory", read_only=False)
conn.register("decimals", decimals)

conn.execute("select * from decimals")
result = conn.df()
```

This yields the error:

```
InvalidInputException: Invalid Input Error: Failed to cast value: Casting value "125.90" to type DECIMAL(38,39) failed: value is out of range!
```

### OS:

macOS, Linux

### DuckDB Version:

0.5.0

### DuckDB Client:

Python

### Full Name:

Dylan Scott

### Affiliation:

Hex Technologies

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #4688

`decimal.Decimal` values leave off the leading zeros in the 'digits' array, this caused the scale to be larger than the width, and thus failing the requirements for the type to become DECIMAL for that value.

In those cases the width is set to scale + 1 (1 to leave room for the non-decimal value)
e.g `0.0001` becomes DECIMAL(5,4)

### CODE REVIEW

**Comment by Tishj** [comment]
The regression test failure looks unrelated / unavoidable to me
These two extra if statements should not impact the performance this much, no?

File: tools/pythonpkg/src/python_objects.cpp [hunk 1]
```diff
@@ -63,6 +63,11 @@ bool PyDecimal::TryGetType(LogicalType &type) {
 		if (exponent_type == PyDecimalExponentType::EXPONENT_POWER) {
 			width += scale;
 		}
+		if (scale > width) {
+			//! The value starts with 1 or more zeros, which are optimized out of the 'digits' array
+			//! 0.001 - width=1, exponent=-3
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:68]: nip : I think //! are only for function descriptions in the headers

**Review by pdet — APPROVED** [review]
LGTM, just small comment

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py [hunk 2]
```diff
@@ -424,6 +425,39 @@ def test_analyze_sample_too_small(self):
...
+                (2502.63),
+                (0.000123),
+                (0.00),
+                (321.00)
+            ) tbl(a);
+        """
+        duckdb_conn.execute(reference_query)
+        reference = duckdb.query("select * from tbl", connection=duckdb_conn).fetchall()
+        conversion = duckdb.query_df(decimals, "x", "select * from x").fetchall()
+
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:458]: Could we also add some tests with more extreme decimal values? e.g.:

* Values that are individually in-range, but out of range when combined: `(0.0000000000000000000000000001), (10000000000000000000000000.0)`
* Mix of very different widths and scales: `(1234.0), (123456789.0), (1234567890123456789.0), (0.1234567890123456789)`
* Values that are out of range entirely: `(1.234567890123456789012345678901234567890), (123456789012345678901234567890123456789.0)`

**Comment by Mytherin** [comment]
Could you still have a go at adding the extra tests? Would love to get this into the bug-fix release next week :)

**Comment by Tishj** [comment]
Apparently these tests just work 👍 
But yea I should have added something similar in advance, the prior tests don't really stress test anything

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py [hunk 3]
```diff
@@ -448,3 +482,76 @@ def test_numeric_decimal_incompatible(self):
...
+            CREATE TABLE tbl AS SELECT * FROM (
+                VALUES
+                (0.0000000000000000000000000001),
+                (10000000000000000000000000.0),
+            ) tbl(a);
+        """
+        duckdb_conn.execute(reference_query)
+        reference = duckdb.query("select * from tbl", connection=duckdb_conn).fetchall()
+        conversion = duckdb.query_df(decimals, "x", "select * from x").fetchall()
+        print(conversion)
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:506]: `assert` instead of `print` maybe? :)

Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:506]: 😅 I thought I had both, and print was there just to see what the result was

Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:510]:
```py
        reference_query = """
            CREATE TABLE tbl AS SELECT * FROM (
                VALUES
                    (1.234567890123456789012345678901234567890),
                    (123456789012345678901234567890123456789.0)
            ) tbl(a);
        """
 reference = duckdb.query("select * from tbl", connection=duckdb_conn).fetchall()
 print(reference)
```
----
```
[(1.2345678901234567,), (1.2345678901234568e+38,)]
```
That's not the object column conversion that's messing up

Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:553]: I've fixed the issues, this time actually asserting ;)
It seems these issues all stem from not supporting conversion to hugeint backed decimals, so I added that and the problems magically disappeared :D

File: tools/pythonpkg/src/python_objects.cpp [hunk 4]
```diff
@@ -63,7 +63,12 @@ bool PyDecimal::TryGetType(LogicalType &type) {
...
 			width += scale;
 		}
-		if (width > Decimal::MAX_WIDTH_INT64) {
+		if (scale > width) {
+			//! The value starts with 1 or more zeros, which are optimized out of the 'digits' array
+			//! 0.001 - width=1, exponent=-3
+			width = scale + 1; // DECIMAL(4,3) - add 1 for the non-decimal values
+		}
+		if (width > Decimal::MAX_WIDTH_INT128) {
 			return false;
```
Comment (Mytherin) [inline tools/pythonpkg/src/python_objects.cpp:72]: Should this perhaps fallback to `DOUBLE`?

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py [hunk 5]
```diff
@@ -490,4 +524,80 @@ def test_numeric_decimal_incompatible(self):
...
+        duckdb_conn.execute(reference_query)
+        reference = duckdb.query("select * from tbl", connection=duckdb_conn).fetchall()
+        conversion = duckdb.query_df(decimals, "x", "select * from x").fetchall()
+        assert(conversion == reference)
+        print(reference)
+        print(conversion)
+
+    def test_numeric_decimal_out_of_range(self):
+        duckdb_conn = duckdb.connect()
+        data = [Decimal("1.234567890123456789012345678901234567"), Decimal("123456789012345678901234567890123456.0")]
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:586]: Python decimals can be arbitrarily large, do we have tests with decimals larger than we can handle somewhere? (e.g. `decimal.Decimal('123456789123456789123456789123456789123456789.5')`)?

**Review by Mytherin — COMMENTED** [review]
Thanks for the fixes! Looks great. One more comment:

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py (hunk 5 above)
Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:586]: Ah that's true, I dont think we do.
The example you sent had more than 38 digits so I truncated it to test the Decimal.
It fell back to VARCHAR, which should indeed be DOUBLE instead.
I'll add a test and the functionality first thing when I come in today :+1:

**Comment by hannes** [comment]
Merging as-is, separate PR for huge decimals @Tishj

**Pull request merged**